cloudformation-telemetry/lambda_function.py

```diff
@@ -0,0 +1,148 @@
+import time
+from crhelper import CfnResource
+from abc import ABC, abstractmethod
+
+
+helper = CfnResource(json_logging=False, log_level='INFO', sleep_on_delete=30)
+
+@helper.create
+def create(event, context):
+    try:
+        T = telemetryFactory(event, context)
+        T.fetch_and_send_telemetry()
+    except Exception as e:
+        print(e)
+        return "Telemetry failed to sent for Create Stack"    
+    helper.Status = "SUCCESS"
+    return "Telemetry sent for Create Stack"
+
+
+@helper.update
+def update(event, context):
+    try:
+        T = telemetryFactory(event, context)
+        T.fetch_and_send_telemetry()
+    except Exception as e:
+        print(e)
+        return "Telemetry failed to sent for Update Stack"    
+    helper.Status = "SUCCESS"
+    return "Telemetry sent for Update Stack"
+
+
+@helper.delete
+def delete(event, context):
+    try:
+        T = telemetryFactory(event, context)
+        T.fetch_and_send_telemetry()
+    except Exception as e:
+        print(e)
+    helper.Status = "SUCCESS"
+
+
+def lambda_handler(event, context):
+    helper(event, context)
+
+def telemetryFactory(event, context):
+    if os.getenv['solutionName'] == 'AWSO':
+        return awsoTelemetry(event, context)
+    else:
+        return None
+    # elif os.environ['solutionName'] == 'CIS':
+    #     return cisTelemetry(event, context)
+
+# Interface
+class baseTelemetry(ABC):
+    @abstractmethod
+    def fetch_and_send_telemetry(self):
+        raise NotImplementedError
+
+    def send_telemetry(self, data):
+        # paramaterise url variable
+        url = 'https://collectors.sumologic.com/receiver/v1/http/ZaVnC4dhaV24CA_LXFO0iHFPLWH8VaEczkwtk-GZYMlTG_Dl2CPQ6YNbmKXf9K3dZQ2aAjTREC_C3TECzVQc1XN7zw5CI5lIR4O4-uYsk4bTELB1MU57AQ=='
+        headers = {'content-type': 'application/json'}
+        print("Telemetry enabled")
+        r = requests.post(url, data = json.dumps(data), headers=headers)
+        
+# class cisTelemetry(baseTelemetry):
+#     def create_telemetry_data(self):
+#         pass
+
+class awsoTelemetry(baseTelemetry):
+
+    def __init__(self, event, context):
+        self.event = event
+        self.context = context
+        self.stackID = os.getenv('stackID')
+        self.cfclient = boto3.client('cloudformation')
+    
+    # This function will return True if any of the child resources are *IN_PROGRESS state.
+    def __has_any_child_resources_in_progress_state(self): # test
+        print("has_any_child function called")
+        all_stacks = self.cfclient.describe_stack_resources(StackName=self.stackID)
+        # PrimeInvoke - only responsible for triggering lambda
+        # Removing 'Primerinvoke' status from all_stacks status so that it is not considered during status checking else it'll result in endless loop becoz if PriveInvoke is not completed overall stack can't be completed.
+        for stack_resource in filter(lambda x: x["LogicalResourceId"] != 'Primerinvoke' ,all_stacks["StackResources"]):
```

Comment:
check if Primerinvoke name is available in context, if CF user gives a different name in their CF then it will fail

Reply:
Implemented